Ionyshch/interface

ionyshch-prismsl opened the pull request.

ionyshch-prismsl committed:
Initial interface version, not all types declared yet

ionyshch-prismsl committed:
Continue describing interface

ionyshch-prismsl committed:
Added example (not yet working), added using Boost

ionyshch-prismsl committed:
Continue defining interface, starting implementation. Added example that should work before SDK is considered for merging into master

ionyshch-prismsl committed:
Added external libraries rapidjson 1.1.0 and easylogging 8.91

ionyshch-prismsl committed:
First GET method working in synchronous mode

ionyshch-prismsl committed:
Separated common and domain types. CurlSession reporesents request-response

ionyshch-prismsl committed:
In progress implementing queryInstrument()

ionyshch-prismsl committed:
First implementation of Metadata, only one level of fields is supported

ionyshch-prismsl committed:
Minimalistik Instrument support, fixed formatting and error handling

ionyshch-prismsl committed:
Implemented uploading event, object stream

ionyshch-prismsl committed:
Implemented all client functions necessary to run example app

ionyshch-prismsl committed:
Example building and running

bvnp44 commented on include/common-types.h:
```diff
@@ -0,0 +1,25 @@
+namespace prism {
+namespace connect {
+
+using std::string;
```
Why we define this in header file? What happen when somebody will include it in another project and write

using namespace prism::connect;
using namespace std;

That using stuff should be used only in *.cpp files.

ionyshch-prismsl replied:
In described scenario everything will be fine.
Please, propose a way to not say std::string each time in header, as we aren't using string other than std::string and it is unlikely we'll do it in future. And if we do, then we can just use fully qualified type e.g. boost::string for *another* string type.

ionyshch-prismsl committed:
Using namespace alias to access members of prism::connect

ionyshch-prismsl committed:
Using namespace alias to hide what chrono is used: boost or std

Yurmusp commented on include/connect.h:
```diff
@@ -5,12 +5,15 @@
+using boost::movelib::unique_ptr;
+using boost::chrono::system_clock;
 
 typedef int32_t accountId_t;
```
I would say it shall be either all capitalized or underscore style.
AccountIdType or account_id_t
I like capitalized more.

Yurmusp commented on src/connect.cpp:
```diff
@@ -0,0 +1,31 @@
+namespace prism {
+namespace connect {
+
+class Client::Impl {
```
Please, do Allman/ANSI style brackets, not K&R

Yurmusp commented on examples/yt_camera/main.cpp:
```diff
@@ -0,0 +1,381 @@
+
+struct CurlGlobal
+{
+    CurlGlobal() { curl_global_init(CURL_GLOBAL_ALL); }
```
Allman/ANSI brackets style needed.

Yurmusp commented on ext/easylogging-8.91/LICENCE.txt:
```diff
@@ -0,0 +1,23 @@
+EasyLogging++ Licence
```
License is written via   "s" , not "c" in the end.

ionyshch-prismsl replied:
Fixed.
I took file as is from that project.

Yurmusp commented on include/client.h:
```diff
@@ -0,0 +1,83 @@
+                         const timestamp_t& startTimestamp, const timestamp_t& stopTimestamp,
+                         const string& videoFile);
+
+    status_t uploadLiveloop(id_t accountId, id_t instrumentId,
```
Liveloop is one word or it shall be LiveLoop?

ionyshch-prismsl replied:
As it is "live loop", I guess LiveLoop is correct. 
Fixed.

Yurmusp commented on include/common-types.h:
```diff
@@ -0,0 +1,25 @@
+using std::vector;
+using std::map;
+
+using boost::movelib::unique_ptr;
```
I do not like "using type;" in headers. 
it becomes a problem later.
it is typically allowed only in *.cpp files.
In headers, I prefer typedef for non templates or
type<T>::t kind of typedef for templates as we had in cpp engine. 
It is still pretty compact and no issues later.

ionyshch-prismsl replied:
Using in header is intentional in particular for boost types, so we don't fully qualify anywhere, but in a single place in the header.
What is the problem with using namespace::template class? It just says when encounter
template class, look for it in namespace. It's controlled and explicit way of introducing names into current namespace.
What will be typedef for vector or map?

Yurmusp replied:
```
template<typename T>
struct vector 
{ 
    typedef std::vector<T> t; 
};
```

Then you use it as
`vector::t<T> vec;`



There is more discussion on it here:
http://stackoverflow.com/questions/6907194/how-to-typedef-a-template-class

Some boost types go to std:: namespace in new compilers.
So, doing this using thing can cause issues with new compilers
if somebody will do using for both: std and prism::connect (or boost::).

Main question here if we want to override allocator for vector, map, string.
If we do want it, then vector::t typedef is way to go.
We had to have it in cpp engine because of two reasons: 
debugging of memory issues and Windows DLLs that may have runtime issues
if allocator is not custom.

If we do not expect overriding allocator, then
Possible way to go can be:
For std::vector and std::map to use fully qualified types with std.
It is not that big of hassle.
If we want to be able to switch to our type later ,then we shall do vector<>::t thing.

For boost::movelib::unique_ptr I would  do unique_ptr<>::t way
or do namespace alias bm::unique_ptr

namespace bm=boost::movelib;
bm::unique_ptr<T> pImpl_;

Otherwise with std::unique_ptr in new compilers is just too easy to get a conflict
when asking to using namespace for both namespaces.

Also, it seems unique_ptr is not really in any public interface call.
Is it right?
I expect it is our implementation internal. So, using short namespace alias for it would be enough.

I thought on it  one way and another and I think 

Most bullet proof way is to use  template typedef like vector::t for all these types.
We can then override allocator, or do whatever needed.
Everything is done in unified way.
It shall also make it able to be DLL on Windows.
We will need to override allocator for that and make vector type different 
from default std::vector.

bvnp44 replied:
Well, we used those things because we had custom memory allocator (as you mentioned), and as for me I don't see any problem for myself to write std::string in header files and define custom namespace simplifications in cpp files

Yurmusp commented on include/const-strings.h:
```diff
@@ -0,0 +1,42 @@
+namespace prism {
+namespace connect {
+
+extern const char* kStrId;
```
Why it starts with k?

ionyshch-prismsl replied:
prefix for const

Yurmusp replied:
Prefix for const seems c .   :)

Yurmusp commented on include/curl-session.h:
```diff
@@ -0,0 +1,130 @@
+    // should they all be pure virtual or with empty implementation, or mix?
+    virtual size_t writeFunction(void* ptr, size_t size, size_t nmemb) = 0;
+    virtual size_t headerFunction(void* ptr, size_t size, size_t nmemb) = 0;
+};
```
Add virtual destructor or base type with virtual destructor

ionyshch-prismsl committed:
Fixes (that doesn't required discussion) based on code review

ionyshch-prismsl committed:
3-clause BSD license added

ionyshch-prismsl committed:
CURL handle pooling implemented, that allows reusing connection and thus reducing connection times. Added dumping CURL performance data

ionyshch-prismsl committed:
Now std::map, std::string and std::vector are referred using fully qualified name (with std::)
Typedef for timestamp_t as number of milliseconds since epoch (time_t(0)), local time.
Fixed toIsoTimeString, so it properly converts milliseconds as %03d (was %d).
Added copyright to header, source and CMakeLists.txt files

Yurmusp commented:
I think using std:: qualifier is good enough compromise.
Only drawback, we can have problems with DLL on Windows,
but we can just not do DLL. We may just stick with leaving everything static lib
and then it shall be no problem.
On Linux SO lib shall have no issue, though. Only Windows have this problem.

Yurmusp commented on CMakeLists.txt:
```diff
@@ -0,0 +1,49 @@
+    src/domain-types.cpp
+    src/util.cpp
+    src/const-strings.cpp
+    examples/yt_camera/main.cpp
```
It is very strange you add main.cpp to static lib.
I guess it is mistake.
In example you create executable and have this main.cpp
included twice. 
Once in static lib and once as main source of executable.
I think it needs to be fixed and it is easy.

ionyshch-prismsl replied:
It's a bug, thanks for noting.

Yurmusp commented on include/domain-types.h:
```diff
@@ -0,0 +1,127 @@
+
+typedef std::vector<CountItem> CountData;
+
+struct EventItem
```
I would name it just Event. 
"Item" is not really needed postfix here. It is considered not good to add such postfixes. They are not really needed and enlongate names and complicate reading/typing.
Same with most other types except some that may cause conflict. 
Please, also fix other names.

ionyshch-prismsl replied:
OK, will be Event for single item and Events for collection

Yurmusp commented on include/domain-types.h:
```diff
@@ -0,0 +1,127 @@
+
+};
+
+typedef std::vector<TagItem> TagData;
```
I think it is not good idea to name vector of items SomethingData.
It shall be either:
Tags . (i like this the most).
or 
TagList or similar (like TagVector).

Because TagData does not imply it is multiple elements container
it sounds identical to TagItem.

bvnp44 commented:
Don't forget to remove QtCreator files :)

ionyshch-prismsl commented:
Why?
Those files are sharable.
Please, provide reason and alternative to solution you don't like.

On Fri, Jan 6, 2017 at 10:09 AM, Yuriy Solovyov <notifications@github.com>
wrote:

> Don't forget to remove QtCreator files :)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/prismskylabs/connect/pull/4#issuecomment-270852941>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AXL5ycZ76hwiKdTp6aQQ7-We1s1FRiHNks5rPfbMgaJpZM4LYOVJ>
> .
>

andrii-prism commented:
CMake is alternative. QTCreator  files imply that QTCreator is required, but in fact is not. SDK should contain only absolute minimum. God forbids we would need to use Windows, then we may add vsproj files.

ionyshch-prismsl commented:
I removed files.

On Fri, Jan 6, 2017 at 10:31 AM, andrii-prism <notifications@github.com>
wrote:

> CMake is alternative.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/prismskylabs/connect/pull/4#issuecomment-270855917>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AXL5ydPodujMKcTjodW3RsTsv2wvJCRiks5rPfv2gaJpZM4LYOVJ>
> .
>
